infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/MultiThreadedThrottlingQueueTest.java

```diff
@@ -13,66 +13,234 @@
   private final MetricTrackingExecutorFactory metricTrackingExecutorFactory =
       new MetricTrackingExecutorFactory(stubMetricsSystem);
   private final DefaultAsyncRunnerFactory asyncRunnerFactory =
       new DefaultAsyncRunnerFactory(metricTrackingExecutorFactory);
-  private final AsyncRunner asyncRunner = asyncRunnerFactory.create("test", 10, 50, 5);
+  private final AsyncRunner asyncRunner = asyncRunnerFactory.create("slow_tasks", 10, 50_000, 5);
+
+  private final AtomicLong rejectedTaskCount = new AtomicLong(0);
 
-  // Primarily a practical demo of tasks interleaving and being limited
-  @Test
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  @Timeout(value = EXPECTED_MAX_TEST_DURATION_MILLIS, unit = TimeUnit.MILLISECONDS)
   @DisabledOnOs(OS.WINDOWS)
-  public void asyncTaskLimitDemo() throws InterruptedException {
-    final ThrottlingTaskQueue queue = new ThrottlingTaskQueue(MAXIMUM_CONCURRENT_TASKS, 10_000);
-    final List<SafeFuture<Void>> futures = new ArrayList<>();
-    for (int i = 0; i < 20; i++) {
-      final int taskNumber = i;
-      futures.add(
-          queue.queueTask(
-              () ->
-                  asyncRunner.runAsync(
-                      () -> {
-                        LOG.info("task {} starting", taskNumber);
-                        int progressCounter = 0;
-                        // 5 x 20 = 100ms at minimum to run
-                        while (progressCounter < 5) {
-                          progressCounter++;
-                          LOG.info("task {}:{}", taskNumber, progressCounter);
-                          Thread.sleep(20);
-                        }
-                        LOG.info("task {} done", taskNumber);
-                      })));
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void heavyMultiThreadedTest(final boolean isPriority) throws InterruptedException {
+    final ThrottlingTaskQueue limitedQueue =
+        isPriority
+            ? new ThrottlingTaskQueueWithPriority(MAXIMUM_CONCURRENT_TASKS, MAXIMUM_QUEUE_SIZE)
+            : new ThrottlingTaskQueue(MAXIMUM_CONCURRENT_TASKS, MAXIMUM_QUEUE_SIZE);
+
+    // We will manually close, ignoring AutoCloseable
+    final ExecutorService producerPool = Executors.newFixedThreadPool(NUM_PRODUCERS);
+
+    // Start all the producer threads
+    for (int i = 0; i < NUM_PRODUCERS; i++) {
+      producerPool.submit(new Producer(limitedQueue, "Producer-" + (i + 1)));
+    }
+
+    // Let the simulation run for the specified duration
+    Thread.sleep(SIMULATION_TIME_SECONDS * 1000L);
+
+    // make sure we had some rejections
+    assertThat(rejectedTaskCount.get()).isGreaterThan(0);
+
+    LOG.info("Simulation time is over. Shutting down threads...");
+
+    // Shutdown the producer pool. They will be interrupted and stop producing.
+    producerPool.shutdownNow();
+
+    try {
+      // Wait for the threads to terminate
+      if (producerPool.awaitTermination(1, TimeUnit.SECONDS)) {
```

Comment:
we could potentially have a global that says the test is running ,and setup the threads to look at that variable to check its still running... would make this less chaotic...